sklearn/pipeline.py

```diff
@@ -37,6 +39,32 @@
 __all__ = ["Pipeline", "FeatureUnion", "make_pipeline", "make_union"]
 
 
+@contextmanager
+def _handle_warnings(estimator):
+    """A context manager to make sure a NotFittedError is raised, if a subestimator
+    raises the error.
+
+    Otherwise, we raise a warning if the pipeline is not fitted, with the deprecation.
+
+    TODO(1.8): remove this context manager and replace with check_is_fitted.
+    """
+    try:
+        yield
+    except NotFittedError:
+        check_is_fitted(estimator)
```

Comment:
If we get here, it means that we raised a `NotFittedError`. Why to we not call `raise` instead of recalling again `check_is_fitted`?

Reply:
Not sure if it answers the question but, given the following estimator:

```python
class MyEstimator:
    @_handle_warning
    def predict(self, X):
        check_is_fitted(self)
        ...
```

When calling `MyEstimator().predict(X)`, the code path will be to go first in `_handle_warning` until the yield and pause there to execute the code from the `predict` method. Then, in the body of `predict` we can raise the `NotFittedError` due to the `check_is_fitted`. This will trigger to go back in `_handle_warning` and continue the code from the `yield` statement afterwards. So we could be in the case that `NotFittedError` was actually raised.

Reply:
Yes, it does answer the question. The only way would actually be to call `check_is_fitted()` somewhere from within the methods. This doesn't happen in the test class, but maybe somewhere else; so I see that now.

Reply:
I'm still wondering why we redo the `check_is_fitted` later :)

Reply:
I now understand that the first 4 lines could cause a `NotFittedError`, because these methods all `return self.steps[-1][1].{method_on_sub-estimator}`.

But why catch the exception just to raise it again, instead of converting it into a warning, which is what I thought we want during the deprecation time...?

Reply:
To answer the original question, because I prefer the `Pipeline` object to raise the `NotFittedError` rather than the exception coming from calling the `predict` of a step instead.

Reply:
But then we could have a `raise NotFittedError(...) from exc`